File: playframework-11506/reproducer-java/app/controllers/HomeController.java
package controllers;

import org.owasp.html.PolicyFactory;
import org.owasp.html.Sanitizers;
import play.cache.Cached;
import play.libs.ws.WSClient;
import play.mvc.*;

import javax.inject.Inject;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.concurrent.CompletionStage;

public class HomeController extends Controller {

    private final WSClient wsClient;

    private int _controllerNumber = 0;
    private int _requestNumber = 0;

    final PolicyFactory policy = Sanitizers.FORMATTING
            .and(Sanitizers.LINKS)
            .and(Sanitizers.TABLES)
            .and(Sanitizers.BLOCKS)
            .and(Sanitizers.IMAGES)
            .and(Sanitizers.STYLES);

    @Inject
    public HomeController(WSClient wsClient) {
        this.wsClient = wsClient;
    }

    @Cached(key="website-sanitized", duration = 20)
    public CompletionStage<Result> websiteSanitized(String address) {
        _controllerNumber += 1;
        System.out.println(_controllerNumber + " controller invocation");
        return this.wsClient.url(address).setRequestTimeout(Duration.of(1000, ChronoUnit.SECONDS)).get()
                .thenApplyAsync(wsResponse -> ok(policy.sanitize(wsResponse.getBody())));
    }

}

File: playframework-11506/reproducer/app/controllers/HomeController.scala
package controllers

import javax.inject._
import play.api._
import play.api.cache.Cached
import play.api.libs.ws.WSClient
import play.api.mvc._
import scala.concurrent.duration._

@Singleton
class HomeController @Inject()
(cached                            : Cached,
 ws                                : WSClient,
 override val controllerComponents : ControllerComponents,
) extends AbstractController(controllerComponents) with Logging {

  import org.owasp.html.Sanitizers

  val policy = Sanitizers.FORMATTING
                         .and(Sanitizers.LINKS)
                         .and(Sanitizers.TABLES)
                         .and(Sanitizers.BLOCKS)
                         .and(Sanitizers.IMAGES)
                         .and(Sanitizers.STYLES)

  def websiteSanitized(address : String)  = cached((reqH: RequestHeader) => f"website-sanitized-${reqH.uri}", duration = 90.seconds) {
    Action.async { implicit request : Request [AnyContent] =>
      //_controllerNumber += 1
      //logger.info(f"> ${_controllerNumber} controller invocation")
      //println(f"> ${_controllerNumber} controller invocation")

      implicit val ec = controllerComponents.executionContext;
      val ok = for {
        response <- {
          //_requestNumber += 1
          //logger.info(f"> ${_requestNumber} Running HTTP request")
          //println(f"> ${_requestNumber} Running HTTP request")

          val upstreamRequest  = ws.url(address).withRequestTimeout(10000.millis)
          upstreamRequest.get()
        }
        result = policy.sanitize(response.body)
      } yield {
        Ok(result)
      }

      ok
    }
  }
}
